pyro/infer/trace_mmd.py

```diff
@@ -0,0 +1,169 @@
+    instead of KL-divergence between variational posterior and prior distributions
+    as in vanilla ELBO. See [1] for the corresponding variant of VAE model,
+    which is a special case of InfoVAE model [2].
+
+    :param kernel: A kernel used to compute MMD.
+        An instance of :class: `pyro.contrib.gp.kernels.kernel.Kernel`,
+        or a dict that maps latent variable names to instances of :class: `pyro.contrib.gp.kernels.kernel.Kernel`.
+        In the latter case, different kernels are used for different latent variables.
+
+    :param mmd_scale: A scaling factor for MMD terms.
+        Float, or a dict that maps latent variable names to floats.
+        In the latter case, different scaling factors are used for different latent variables.
+
+    References
+
+    [1] `A Tutorial on Information Maximizing Variational Autoencoders (InfoVAE)`
+        Shengjia Zhao
+        https://ermongroup.github.io/blog/a-tutorial-on-mmd-variational-autoencoders/
+
+    [2] `InfoVAE: Balancing Learning and Inference in Variational Autoencoders`
+        Shengjia Zhao, Jiaming Song, Stefano Ermon
+    """
+
+    def __init__(self,
+                 kernel, mmd_scale=1,
+                 num_particles=1,
+                 max_plate_nesting=float('inf'),
+                 max_iarange_nesting=None,  # DEPRECATED
+                 vectorize_particles=False,
+                 strict_enumeration_warning=True,
+                 ignore_jit_warnings=False,
+                 retain_graph=None):
+        super().__init__(
+            num_particles, max_plate_nesting, max_iarange_nesting, vectorize_particles,
+            strict_enumeration_warning, ignore_jit_warnings, retain_graph,
+        )
+        self._kernel = None
+        self._mmd_scale = None
+        self.kernel = kernel
+        self.mmd_scale = mmd_scale
+
+    @property
+    def kernel(self):
+        return self._kernel
+
+    @kernel.setter
+    def kernel(self, kernel):
+        if isinstance(kernel, dict):
+            self._kernel = kernel
+        elif isinstance(kernel, pyro.contrib.gp.kernels.kernel.Kernel):
+            self._kernel = defaultdict(lambda: kernel)
+        else:
+            raise TypeError("`kernel` should be an instance of `pyro.contrib.gp.kernels.kernel.Kernel`")
+
+    @property
+    def mmd_scale(self):
+        return self._mmd_scale
+
+    @mmd_scale.setter
+    def mmd_scale(self, mmd_scale):
+        if isinstance(mmd_scale, dict):
+            self._mmd_scale = mmd_scale
+        elif isinstance(mmd_scale, (int, float)):
+            self._mmd_scale = defaultdict(lambda: float(mmd_scale))
+        else:
+            raise TypeError("`mmd_scale` should be either float, or a dict of floats")
+
+    def _differentiable_loss_parts(self, model, guide, *args, **kwargs):
+        loglikelihood = 0.0
+        penalty = 0.0
+        model_trace_independent = poutine.trace(model, graph_type="flat").get_trace(*args, **kwargs)
```

Comment:
This will not have the correct number of particles when `self.num_particles != 1`.  You can use `ELBO._vectorized_num_particles` to handle that:
```
model_trace_independent = poutine.trace(self._vectorized_num_particles(model)).get_trace(*args, **kwargs)
```

Reply:
You are right, I didn't notice the bug, because I didn't test with `vectorize_particles=True`